File: apps/backend-gateway/src/infrastructure/security/rbac.ts
/**
 * StadiumOS AI — Authentication & Role-Based Access Control Middlewares.
 * 
 * Verifies JWT session headers and asserts RBAC role levels hierarchies.
 */

import { Request, Response, NextFunction } from "express";
import { UserRole } from "../../domain/enums";
import { AuthenticationError, AuthorizationError } from "../../domain/exceptions";
import { verifyToken } from "./securityHelpers";
import { prisma } from "../database/prisma";
import { User } from "../../domain/entities/user";

export interface AuthenticatedRequest extends Request {
  user?: User;
}

/**
 * Express middleware to extract JWT Bearer token and verify User profile existence.
 */
export async function authenticateMiddleware(
  req: AuthenticatedRequest,
  res: Response,
  next: NextFunction
): Promise<void> {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith("Bearer ")) {
    next(new AuthenticationError("Authorization header is missing or malformed."));
    return;
  }

  const token = authHeader.split(" ")[1];
  if (!token) {
    next(new AuthenticationError("Bearer token credentials missing."));
    return;
  }

  try {
    const payload = verifyToken(token, "access");
    const userId = payload.sub;

    if (!userId) {
      throw new AuthenticationError("Invalid access token structure.");
    }

    // Load active profile from database
    const userModel = await prisma.user.findUnique({
      where: { id: userId }
    });

    if (!userModel) {
      throw new AuthenticationError("Authenticated user account was not found.");
    }

    if (!userModel.is_active) {
      throw new AuthenticationError("Authenticated user account is deactivated.");
    }

    // Map database model to domain Entity instance
    req.user = new User(
      userModel.id,
      userModel.email,
      userModel.hashed_password,
      userModel.role as UserRole,
      userModel.is_active,
      userModel.created_at,
      userModel.updated_at
    );

    next();
  } catch (err) {
    next(err);
  }
}

/**
 * Parametric middleware factory enforcing minimum RBAC authorization level.
 */
export function requireRole(requiredRole: UserRole) {
  return (req: AuthenticatedRequest, res: Response, next: NextFunction): void => {
    if (!req.user) {
      next(new AuthenticationError("User session context is unauthenticated."));
      return;
    }

    if (!req.user.hasPermission(requiredRole)) {
      next(new AuthorizationError(requiredRole, req.user.role));
      return;
    }

    next();
  };
}

File: apps/backend-gateway/src/infrastructure/security/rbac.py
"""
StadiumOS AI — Authentication & Role-Based Access Control Dependencies.

FastAPI security dependencies to extract user context from JWT tokens and enforce
RBAC requirements on endpoint access.
"""

from typing import Annotated
from uuid import UUID
from fastapi import Depends
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
from sqlalchemy.ext.asyncio import AsyncSession

from src.domain.enums import UserRole
from src.domain.exceptions import AuthenticationError, AuthorizationError
from src.infrastructure.database.engine import get_db_session
from src.infrastructure.repositories.user_repository import SqlUserRepository
from src.infrastructure.security.jwt_handler import verify_token
from src.domain.entities.user import User

security_scheme = HTTPBearer(auto_error=False)


async def get_current_user(
    credentials: Annotated[HTTPAuthorizationCredentials | None, Depends(security_scheme)],
    db: Annotated[AsyncSession, Depends(get_db_session)],
) -> User:
    """
    Extract the Bearer JWT token from request headers, verify it, and load
    the associated user entity.

    Raises:
        AuthenticationError: If token is missing, invalid, or expired.
    """
    if not credentials:
        raise AuthenticationError("Authorization header is missing.")

    token = credentials.credentials
    # Validates access signature and exp
    payload = verify_token(token, expected_type="access")
    
    user_id_str = payload.get("sub")
    if not user_id_str:
        raise AuthenticationError("Invalid access token structure.")

    user_repo = SqlUserRepository(db)
    user = await user_repo.get_by_id(UUID(user_id_str))
    if not user:
        raise AuthenticationError("Authenticated user account was not found.")

    if not user.is_active:
        raise AuthenticationError("Authenticated user account is deactivated.")

    return user


class RequireRole:
    """
    Dependency that restricts endpoint access to users possessing a minimum RBAC role level.

    The hierarchy is checked using User.has_permission(role).
    """

    def __init__(self, required_role: UserRole) -> None:
        self.required_role = required_role

    def __call__(self, current_user: Annotated[User, Depends(get_current_user)]) -> User:
        """
        Evaluate user permissions.

        Raises:
            AuthorizationError: If the user's role is below the required role level.
        """
        if not current_user.has_permission(self.required_role):
            raise AuthorizationError(
                required_role=self.required_role.value,
                current_role=current_user.role.value,
            )
        return current_user
